### PR TITLE
[WIP] Integrate WDIO 

### PR DESCRIPTION
**Pull request by conradkirschner**
Reason behind that pullrequest is, to integrate wdio. This will allow BackstopJS to use selenium/webdriver standards. 

- [ ] define how config should look like
- [ ] check if cloud provider will work 
- [ ] do just the minimal changes for now -> make it run with wdio and exclude features that are not possible (throw warning)

Currently the branch is Work in Progress, but feel free to have a first look.

### CODE REVIEW

**Comment by conradkirschner** [comment]
waiting for feedback. 

current blocker is settting window size ....

**Comment by garris** [comment]
Wow.  This looks like a lot of work.  I think I'll add a `/wdio` branch.   Will try get to this soon.

**Comment by conradkirschner** [comment]
not that much work (1 evening + 1 day) I just tried to replace puppeteer everwhere with wdio. so incremental exchange commands 

puppeteer: 
page.evaluate()

wdio:
browser.execute();
 
 copied the engine scripts and converted them aswell ... 

(I need to spend a bit more time on it, before I would remove the WIP - like the window size height thing is most confusing me)

But sure I will try to finish that pull request soon, but I like to make my work transparent (therefore WIP)

**Comment by conradkirschner** [comment]
Some features are missing in wdio standalone, that should not be part of that project. 
https://github.com/webdriverio/webdriverio/issues/7185

**Comment by conradkirschner** [comment]
As suggested in webdriverio/webdriverio#7185 this should not be part of wdio core, so I added service handling directly in backstopjs, need to verify if that solution works (just tried out with selenium-standalone-service)

**Review by garris — COMMENTED** [review]
Left some comments -- thanks for working on this!

File: README.md [hunk 1]
```diff
@@ -29,6 +29,9 @@
 - Plays nice with CI and source control
 - Run globally or locally as a standalone package app or `require('backstopjs')` right into your node app
 - Incredibly easy to use: just 3 commands go a long long way!
+- [BETA] Limited Support for WebDriver Protocol, based on wdio
+   - works as well with any wdio-(cloud)-service (see [wdio configuration]() )
+   - not compatible with puppeteer references pictures
```
[the remark this answers is not in the document]
Comment (conradkirschner) [reply README.md:34]: I had some issues with directly comparing, backstopjs puppeteer pictures with wdio pictures. 
- screensize is different a bit (still need to investigate if that what goes on, but due sickness&work I couldn't find the time) 
- puppeteer does a full page picture, webdriver captures only viewport => In theory, should be possible to merge the viewport images to a full page picture, wdio plugin is outdated
- image interception is not possible without a proxy solution, due selenium limitation 


Therefore I added a BETA support, so maybe some people get interessted to work on it.

[the remark this answers is not in the document]
Comment (garris) [reply README.md:34]: Oh, it is very very unlikely that you could ever render a reference image in puppeteer and then be able to test that against a test image generated in WDIO.  Or vice versa.   In addition to different high level rendering interpretations of the different application environments, there will also be a low-level rendering issue because different hardware will dither bitmap text differently.  [For example you can see this explanation](https://github.com/garris/BackstopJS#using-docker-for-testing-across-different-environments).    

TLDR;   Screenshots generated by a single hardware, software and platform combination can only be compared against itself.

**Comment by conradkirschner** [comment]
@garris  Regarding the different environments, I just added it into the readme to warn people that cross engine will not work. 
https://github.com/garris/BackstopJS/pull/1342#discussion_r689730423

But as it is already written in the readme, might be better as a own section... 
(removed it for now)
